### PR TITLE
[CALCITE-3174] IS NOT DISTINCT FROM condition pushed from filter to join is not collapsed

### PR DESCRIPTION
**Pull request by KazydubB**

### CODE REVIEW

**Comment by rubenada** [comment]
LGTM

**Comment by jinxing64** [comment]
@KazydubB THX for this PR and it really make sense for me.
 Could you please add some short description above ?

File: core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java [hunk 1]
```diff
@@ -275,6 +276,28 @@ protected void perform(RelOptRuleCall call, Filter filter,
...
+  /**
+   * Get conjunctions of filter's condition but with collapsed
+   * {@code IS NOT DISTINCT FROM} expressions if needed.
+   *
+   * @param filter filter containing condition
+   * @return condition conjunctions with collapsed {@code IS NOT DISTINCT FROM}
+   * expressions if any
+   * @see RelOptUtil#conjunctions(RexNode)
+   */
+  private List<RexNode> getConjunctions(Filter filter) {
```
Comment (jinxing64) [inline core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java]: For the Join condition, it seems like we should always remember for `collapseExpandedIsNotDistinctFromExpr`. If new rule added, users should do this expansion again, just like this PR and https://github.com/apache/calcite/blob/master/core/src/main/java/org/apache/calcite/rel/rules/ProjectJoinTransposeRule.java#L92
I mean that shall we extract such behavior as a shared `RexShuttle`, and have an approach to make this expansion automatic ?

File: core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java [hunk 2]
```diff
@@ -275,6 +276,28 @@ protected void perform(RelOptRuleCall call, Filter filter,
...
+    List<RexNode> conjunctions = conjunctions(filter.getCondition());
+    RexBuilder rexBuilder = filter.getCluster().getRexBuilder();
+    for (int i = 0; i < conjunctions.size(); i++) {
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java]: final RexBuilder rexBuilder

File: core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java (hunk 1 above)
Comment (danny0405) [reply core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java]: This returns conjunctions of filer condition, and if only 2 places use this code, there is no necessity to have a common `RexShuttle`.

**Pull request closed without merging**